quickwit-doc-mapper/src/routing_expression/mod.rs

```diff
@@ -0,0 +1,171 @@
+
+pub trait RoutingExprContext {
+    // TODO see if we can get rid of the alloc in some specific case
+    fn hash_attribute<H: Hasher>(&self, attr_name: &str, hasher: &mut H);
+}
+
+/// This is a bit overkill but this function has the merit of
+/// ensuring that the data that is sent to the hasher is unique
+/// to the value, so we do not lose injectivity there.
+fn hash_json_val<H: Hasher>(json_val: &serde_json::Value, hasher: &mut H) {
+    match json_val {
+        serde_json::Value::Null => {
+            hasher.write_u8(0u8);
+        }
+        serde_json::Value::Bool(bool_val) => {
+            hasher.write_u8(1u8);
+            bool_val.hash(hasher);
+        }
+        serde_json::Value::Number(num) => {
+            hasher.write_u8(2u8);
+            num.hash(hasher);
+        }
+        serde_json::Value::String(s) => {
+            hasher.write_u8(3u8);
+            hasher.write_usize(s.len());
+            hasher.write(s.as_bytes());
+        }
+        serde_json::Value::Array(arr) => {
+            hasher.write_u8(4u8);
+            hasher.write_usize(arr.len());
+            for el in arr {
+                hash_json_val(el, hasher);
+            }
+        }
+        serde_json::Value::Object(obj) => {
+            hasher.write_u8(5u8);
+            hasher.write_usize(obj.len());
+            for (key, val) in obj.iter() {
+                hasher.write_usize(key.len());
+                hasher.write(key.as_bytes());
+                hash_json_val(val, hasher);
+            }
+        }
+    }
+}
+
+impl RoutingExprContext for serde_json::Map<String, serde_json::Value> {
+    fn hash_attribute<H: Hasher>(&self, attr_name: &str, hasher: &mut H) {
+        if let Some(json_val) = self.get(attr_name) {
+            hasher.write_u8(1u8);
+            hash_json_val(json_val, hasher);
+        } else {
+            hasher.write_u8(0u8);
+        }
+    }
+}
+
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub enum RoutingExpr {
+    Attribute(String),
+    List(Vec<RoutingExpr>),
+    // TODO Enrich me! Map / Modulo
+}
+
+impl FromStr for RoutingExpr {
+    type Err = anyhow::Error;
+
+    fn from_str(expr_dsl_str: &str) -> anyhow::Result<Self> {
+        if expr_dsl_str.is_empty() {
+            return Ok(RoutingExpr::List(Vec::new()));
+        }
+        Ok(RoutingExpr::Attribute(expr_dsl_str.to_string()))
+    }
+}
+
+impl ToString for RoutingExpr {
+    fn to_string(&self) -> String {
+        match self {
+            RoutingExpr::Attribute(field) => field.to_string(),
+            RoutingExpr::List(children) => children.iter().map(ToString::to_string).join(","),
+        }
+    }
+}
+
+#[derive(Hash)]
+#[repr(u8)]
+enum ExprType {
+    Leaf,
+    List,
+}
+
+impl RoutingExpr {
+    /// Evaluate the expression applied to the given
+    /// context and returns a u64 hash.
+    ///
+    /// Obviously this function is not perfectly injective.
+    pub fn eval_hash<Ctx: RoutingExprContext>(&self, ctx: &Ctx) -> u64 {
```

Comment:
What is the value returned for the empty expression `List([])`? 0?

Reply:
No. It is the hash of some short payload. It does not really matter though.